File: crates/fey_math/src/affine3.rs
use crate::{Affine2, Angle, Float, Mat3, Num, Vec2, Vec3, impl_affine};

pub type Affine3F = Affine3<f32>;

/// A 3D affine matrix (translation, rotation, and scaling).
#[repr(C)]
#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct Affine3<T> {
    pub matrix: Mat3<T>,
    pub translation: Vec3<T>,
}

impl_affine!(
    NAME = Affine3
    SHORT = affine3
    MUL_FN = mul_affine3
    MAT_TY = Mat3
    VEC_TY = Vec3
);

/// Create an [`Affine3`].
#[inline]
pub const fn affine3<T>(matrix: Mat3<T>, translation: Vec3<T>) -> Affine3<T> {
    Affine3 {
        matrix,
        translation,
    }
}

impl<T: Float> Affine3<T> {
    /// Create a rotation matrix along the x-axis.
    #[inline]
    pub fn rotation_x(angle: impl Angle<T>) -> Self {
        affine3(Mat3::rotation_x(angle), Vec3::ZERO)
    }

    /// Create a rotation matrix along the y-axis.
    #[inline]
    pub fn rotation_y(angle: impl Angle<T>) -> Self {
        affine3(Mat3::rotation_y(angle), Vec3::ZERO)
    }

    /// Create a rotation matrix along the z-axis.
    #[inline]
    pub fn rotation_z(angle: impl Angle<T>) -> Self {
        affine3(Mat3::rotation_z(angle), Vec3::ZERO)
    }

    /// Transforms a 2D vector.
    #[inline]
    pub fn transform_vec2(&self, rhs: Vec2<T>) -> Vec2<T> {
        self.matrix.transform_vec2(rhs)
    }

    /// Transforms a 2D point.
    #[inline]
    pub fn transform_pos2(&self, rhs: Vec2<T>) -> Vec2<T> {
        self.matrix.transform_pos2(rhs) + self.translation.xy()
    }

    /// Transforms a 3D vector.
    #[inline]
    pub fn transform_vec3(&self, rhs: Vec3<T>) -> Vec3<T> {
        self.matrix.transform_vec3(rhs)
    }

    /// Transforms a 3D point.
    #[inline]
    pub fn transform_pos3(&self, rhs: Vec3<T>) -> Vec3<T> {
        self.matrix.transform_vec3(rhs) + self.translation
    }

    /// Try to invert the matrix.
    #[inline]
    pub fn inverse(&self) -> Option<Self> {
        let matrix = self.matrix.inverse()?;
        let translation = -matrix.transform_vec3(self.translation);
        Some(affine3(matrix, translation))
    }

    /// Multiply by another matrix.
    #[inline]
    pub fn mul_affine3(&self, rhs: &Self) -> Self {
        affine3(
            self.matrix * rhs.matrix,
            self.matrix.transform_vec3(rhs.translation) + self.translation,
        )
    }
}

impl<T: Num> From<Affine2<T>> for Affine3<T> {
    #[inline]
    fn from(
        Affine2 {
            matrix,
            translation,
        }: Affine2<T>,
    ) -> Self {
        affine3(matrix.into(), translation.into())
    }
}

File: crates/fey_math/lua/Affine3.lua
---@meta

---A 3D affine matrix (translation, rotation, scaling and shear).
---@class (exact) Affine3: Affine3Methods
---@field matrix Mat3
---@field translation Vec3
---@operator mul(Affine3): Affine3

---@class Affine3Class : Affine3Methods
---@overload fun(matrix: Mat3, translation: Vec3): Affine3
local module = {}

---@class Affine3Methods
local methods = {}

---An identity matrix.
---@return Affine3
---@nodiscard
function module.identity() end

---A zero matrix.
---@return Affine3
---@nodiscard
function module.zero() end

---Create a new matrix.
---@param matrix Mat3
---@param translation Vec3
---@return Affine3
---@nodiscard
function module.new(matrix, translation) end

---Create a matrix rotating around the x-axis.
---@param angle number Angle (in radians).
---@return Affine3
---@nodiscard
function module.rotation_x(angle) end

---Create a matrix rotating around the y-axis.
---@param angle number Angle (in radians).
---@return Affine3
---@nodiscard
function module.rotation_y(angle) end

---Create a matrix rotating around the z-axis.
---@param angle number Angle (in radians).
---@return Affine3
---@nodiscard
function module.rotation_z(angle) end

---Create a scaling matrix.
---@param scale Vec3|number
---@return Affine3
---@nodiscard
function module.scale(scale) end

---Create a translation matrix.
---@param translation Vec3
---@return Affine3
---@nodiscard
function module.translation(translation) end

---Returns a temporary copy of this value.
---@param self Affine3
---@return Affine3
---@nodiscard
function methods.clone(self) end

---Boxes the value if not already boxed.
---@param self Affine3
---@return Affine3
---@nodiscard
function methods.box(self) end

---Boxes a copy of this value.
---@param self Affine3
---@return Affine3
---@nodiscard
function methods.box_clone(self) end

---Returns `true` if the matrices are approximately equal.
---@param self Affine3
---@param other Affine3
---@return boolean
---@nodiscard
function methods.approx(self, other) end

---Inverts the matrix.
---@param self Affine3
---@return Affine3?
---@nodiscard
function methods.inverse(self) end

---Multiplies the two matrices.
---@param self Affine3
---@param other Affine3
---@return Affine3
---@nodiscard
function methods.mul_affine3(self, other) end

---Transform the 2D vector by this matrix.
---@param self Affine3
---@param vec Vec2
---@return Vec2
---@nodiscard
function methods.transform_vec2(self, vec) end

---Transform the 2D point by this matrix.
---@param self Affine3
---@param pos Vec2
---@return Vec2
---@nodiscard
function methods.transform_pos2(self, pos) end

---Transform the 3D vector by this matrix.
---@param self Affine3
---@param vec Vec3
---@return Vec3
---@nodiscard
function methods.transform_vec3(self, vec) end

---Transform the 3D point by this matrix.
---@param self Affine3
---@param pos Vec3
---@return Vec3
---@nodiscard
function methods.transform_pos3(self, pos) end

return module